### PR TITLE
Enhance SQLNodeConverterEngine to support 'select_string_constant_type_cast'

### PR DESCRIPTION
**Pull request by cardigan1008**
Ref #26871 .

Changes proposed in this pull request:
  - Finish task 5: select_string_constant_type_cast
  - dbtype:PostgreSQL, openGauss

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: test/it/optimizer/src/test/resources/converter/select.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<sql-node-converter-test-cases>
+    <test-cases sql-case-id="select_string_constant_type_cast" expected-sql="SELECT CAST('1' AS INTEGER), CAST('2' AS DECIMAL)" dbtypes="openGauss,PostgreSQL" />
+</sql-node-converter-test-cases>
```
Comment (strongduanmu) [inline test/it/optimizer/src/test/resources/converter/select.xml:21]: Please add new line at the end of file.

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java [hunk 2]
```diff
@@ -43,7 +43,25 @@ public Optional<SqlNode> convert(final TypeCastExpression segment) {
         if (!expression.isPresent()) {
             return Optional.empty();
         }
-        SqlTypeNameSpec sqlTypeName = new SqlBasicTypeNameSpec(SqlTypeName.valueOf(segment.getDataType().toUpperCase()), SqlParserPos.ZERO);
+        String typeName = segment.getDataType().toUpperCase();
+        switch (typeName) {
+            case "INT2":
```
Comment (strongduanmu) [inline kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java:48]: Why do this type convert?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java (hunk 2 above)
Comment (cardigan1008) [reply kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java:48]: > Why do this type convert?

You can view the supported SQL data types in org/apache/calcite/sql/type/SqlTypeName.java. This file contains the enumeration of all the standard SQL data types that are supported by Calcite. If you want to use a non-standard SQL data type in your queries, you will need to map it to its closest standard SQL data type before passing it to Calcite.

Comment (strongduanmu) [reply kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java:48]: > > Why do this type convert?
> 
> You can view the supported SQL data types in org/apache/calcite/sql/type/SqlTypeName.java. This file contains the enumeration of all the standard SQL data types that are supported by Calcite. If you want to use a non-standard SQL data type in your queries, you will need to map it to its closest standard SQL data type before passing it to Calcite.

Ok, I got it. Is this conversion logic only valid for PostgreSQL? We should probably design the SPI by databaseType to isolate these logics.

Comment (cardigan1008) [reply kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java:48]: > > > Why do this type convert?
> > 
> > 
> > You can view the supported SQL data types in org/apache/calcite/sql/type/SqlTypeName.java. This file contains the enumeration of all the standard SQL data types that are supported by Calcite. If you want to use a non-standard SQL data type in your queries, you will need to map it to its closest standard SQL data type before passing it to Calcite.
> 
> Ok, I got it. Is this conversion logic only valid for PostgreSQL? We should probably design the SPI by databaseType to isolate these logics.

Not really, the same conversion issue may arise when using OpenGauss with Calcite.

Comment (strongduanmu) [reply kernel/sql-federation/core/src/main/java/org/apache/shardingsphere/sqlfederation/compiler/converter/segment/expression/impl/TypeCastExpressionConverter.java:48]: openGauss is a derivative version of PostgreSQL database, these types do not exist in sqlserver, oracle, so they are not common logic.

**Comment by strongduanmu** [comment]
Hi @cardigan1008, any progress for this pr?

**Comment by cardigan1008** [comment]
Hi @strongduanmu , I have implemented the DataTypeConverter to handle the data type conversions. Please review the changes.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**